pkg/transform/client.go

```diff
@@ -0,0 +1,306 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/bufbuild/buf/pkg/transform/internal/cache"
+	"github.com/bufbuild/buf/private/gen/proto/connect/buf/alpha/registry/v1alpha1/registryv1alpha1connect"
+	registryv1alpha1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/registry/v1alpha1"
+	"github.com/bufbuild/connect-go"
+)
+
+// Client ..
+type Client struct {
+	client                 registryv1alpha1connect.SchemaServiceClient
+	cache                  *cache.Cache
+	owner                  string
+	repository             string
+	version                string
+	types                  []string
+	discardUnknown         bool
+	inputFormat            registryv1alpha1.Format
+	ifNotCommit            string
+	excludeCustomOptions   bool
+	excludeKnownExtensions bool
+	outputFormat           interface{}
+}
+
+type ClientBuilder func(*Client)
+
+// NewClient builds a transform client, processing various configurable options
+func NewClient(ctx context.Context, builders ...ClientBuilder) (*Client, error) {
+	client := &Client{}
+	for _, apply := range builders {
+		apply(client)
+	}
+	if err := client.validate(); err != nil {
+		return nil, err
+	}
+	if client.cache != nil {
+		// pre-load proto encoding resolver into cache
+		if _, err := client.getProtoEncodingResolver(ctx); err != nil {
+			return nil, err
+		}
+	}
+	return client, nil
+}
+
+func DefaultClient(
+	ctx context.Context,
+	owner, repository, version string,
+	inputFormat registryv1alpha1.Format,
+	outputFormat ClientBuilder,
+) (*Client, error) {
+	return NewClient(
+		ctx,
+		WithDefaultSchemaService(),
+		WithDefaultCache(),
+		WithBufModule(owner, repository, version),
+		FromFormat(inputFormat, false),
+		outputFormat,
+	)
+}
+
+// WithNewSchemaService configures the remote Buf Schema Registry. Accepting a
+// HTTPClient, baseURL and connect client options.
+func WithNewSchemaService(
+	httpClient connect.HTTPClient,
+	baseURL string,
+	opts ...connect.ClientOption,
+) ClientBuilder {
+	return func(c *Client) {
+		c.client = registryv1alpha1connect.NewSchemaServiceClient(
+			httpClient,
+			baseURL,
+			opts...,
+		)
+	}
+}
+
+func WithDefaultSchemaService() ClientBuilder {
+	return WithNewSchemaService(http.DefaultClient, "buf.build")
+}
+
+func WithSchemaService(
+	client registryv1alpha1connect.SchemaServiceHandler,
+) ClientBuilder {
+	return func(c *Client) {
+		c.client = client
+	}
+}
+
+// WithCache provided a default expiration duration and cleanup
+// interval. If the expiration duration is less than one (or NoExpiration),
+// the items in the cache never expire (by default), and must be deleted
+// manually. If the cleanup interval is less than one, expired items are not
+// deleted from the cache before calling c.DeleteExpired().
+func WithCache() ClientBuilder {
```

Comment:
I think we'd want the caller to provide the cache, no? `WithDefaultCache` below could use a simple in-memory cache. But caller could provide some impl that is backed by redis or memcache, etc. The cache type could be a simple interface like:
```go
type SchemaCache interface {
    Load() (schema *descriptorpb.FileDescriptorSet, commit string, error)
    Save(schema *descriptorpb.FileDescriptorSet, commit string) error
}
```